ebr.go

```
@@ -0,0 +1,251 @@
+// Package the rules logic
+// The exported function ApplyOn check if requiered conditions are valid and execute actions from entityRules object
+
+package entityrules
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/go-bexpr"
+	"reflect"
+	"regexp"
+	"strings"
+	"text/template"
+)
+
+type EntityRules struct {
+	Conditions *MethodsExecution
+	Actions    *MethodsExecution
+	Rules      Rules
+}
+type MethodsExecution []MethodExecution
+
+type MethodExecution struct {
+	Name      string
+	Method    string
+	Arguments Arguments
+}
+type Arguments []Argument
+type Argument interface{}
+
+type Rules []Rule
+
+type Rule struct {
+	Name        string
+	Description string
+	When        []string
+	If          string
+	Do          []string
+}
+
+var (
+	ErrNotRegistered = errors.New("is not registered")
+	ErrIsNotBool     = errors.New("do not return boolean")
+)
+
+func (entityRules *EntityRules) ApplyOn(entities ...interface{}) error {
+	if err := entityRules.Validate(); err != nil {
+		return err
+	}
+	for _, e := range entities {
+		if err := entityRules.applyOnOne(e); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (entityRules *EntityRules) Validate() error {
+	if err := entityRules.Actions.validate(); err != nil {
+		return err
+	}
+	if err := entityRules.Conditions.validate(); err != nil {
+		return err
+	}
+	if rr := entityRules.Rules; rr == nil {
+		fmt.Println("Entity does not have rules")
```

remove this

---

you should return an error if no rules